Public/Src/Cache/ContentStore/Distributed/NuCache/RaidedRedisDatabase.cs

```diff
@@ -101,8 +106,44 @@ public async Task<(TResult primary, TResult secondary)> ExecuteRaidedAsync<TResu
                 await primaryResultTask.IgnoreErrorsAndReturnCompletion();
             }
 
-            var secondaryResultTask = ExecuteAndCaptureRedisErrorsAsync(SecondaryRedisDb, executeAsync);
-            await Task.WhenAll(primaryResultTask, secondaryResultTask);
+            var secondaryResultTask = ExecuteAndCaptureRedisErrorsAsync(SecondaryRedisDb, executeAsync, cancellationTokenSource.Token);
+
+            // Instead of waiting for both - the primary and the secondary, we'll check for the primary first and then try to cancel the other one.
+            // There is a time out delay acting as a window for the slower task to complete before we cancel the retry attempts
+
+            Task<TResult> fasterResultTask = await Task.WhenAny(primaryResultTask, secondaryResultTask);
+            Task<TResult> slowerResultTask = fasterResultTask == primaryResultTask ? secondaryResultTask : primaryResultTask;
+
+            // Try to cancel the slower operation only when the faster one finished successfully (and the timeout was provided).
+            if (fasterResultTask.Result.Succeeded && retryWindow != null)
+            {
+                // Giving the second task a chance to succeed.
+                Task secondResult = await Task.WhenAny(slowerResultTask, Task.Delay(retryWindow.Value, context.Token));
+
+                if (secondResult != slowerResultTask)
+                {
+                    // Avoiding task unobserved exception if the second task will fail.
+                    // TODO ST: pass the severity into this operation to trace the failure as a debug severity message (not available yet).
+                    slowerResultTask.FireAndForget(context);
+
+                    // The second task is not done within a given timeout.
+                    cancellationTokenSource.Cancel();
+
+                    var failingRedisDb = GetDbName(fasterResultTask == primaryResultTask ? SecondaryRedisDb : PrimaryRedisDb);
+                    Tracer.Info(context, $"{Tracer.Name}.{caller}: Error in {failingRedisDb} redis db using result from other redis db: {fasterResultTask}");
+
+                    if (fasterResultTask == primaryResultTask)
+                    {
+                        return (await fasterResultTask, default);
+                    }
+                    else
+                    {
+                        return (default, await fasterResultTask);
+                    }

```

Comment:
This is not safe. The '|' (or) operator for BoolResult is not null safe, so returning default will cause a null reference exception.